pkg/kafka/policy.go

```diff
@@ -15,13 +15,27 @@
 package kafka
 
 import (
-	"github.com/cilium/cilium/pkg/flowdebug"
 	"github.com/cilium/cilium/pkg/policy/api"
 
 	"github.com/optiopay/kafka/proto"
-	"github.com/sirupsen/logrus"
 )
 
+func matchNonTopicRequests(req *RequestMessage, rule api.PortRuleKafka) bool {
+	// These are no topic requests. If topic is present, we need to return
+	// false since topic can never be associated with this request kind
+	// The topic can be non null if APIRole is specified.
+	// If topic is non null, and so is APIKey, this is due to the policy
+	// correlating a non-topic apikey like apiversions with topic.
+	// We should return false in this case.
+	if rule.Topic != "" && len(rule.APIKey) != 0 {
```

Comment:
Shouldn't we always return `false` regardless of whether APIKey is specified? The rule can't possibly match as we can't match on the topic.

Reply:
Actually it can. If we specify something like
```
"kafka": [
{"Role": "produce", "topic": "allowedTopic"},
{"Role": "consume", "topic": "allowedTopic"}
]
```
versus 
something like 
```
"kafka": [
{"apiKey": "produce", "topic": "allowedTopic"},
{"apiKey": "apiVersions}
]
```

The first case will expand it to include all relevant keys for Role "produce", coupled with the associated "topic"

Please let me know if you think this is incorrect and we can discuss this further.

Reply:
I'm trying to understand the following logic:
`if rule.Topic != "" && len(rule.APIKey) != 0 {`

`matchNonTopicRequests()` is called when the kafka parser was not able to parse beyond the generic header. So there is no possibility that we can know whether the the APIkey specific header even includes a topic and if so, what the value is, therefore I don't understand how a rule that requires a topic to be of a particular value could ever match. Why should the `return false` be dependant on whether an APIkey was specified as well?

Reply:
Well we check earlier if only one of the two i.e` rule.APIKey `or `rule.Role` is set.
If `rule.APIRole` is set, it is expanded to include all the apiKeys. 

This code will not be reached if rule.APIKey is not set, but it can be reached if rule.APIRole is not nil
Basically this can be rewritten as 
```
if rule.Topic != "" && len(rule.Role) == 0
```

Reply:
I think `if rule.Topic != "" && len(rule.Role) == 0` is a more cleaner way to specify this. Do you agree?

Reply:
Is the intent to allow requests which have no topic field or which we don't understand beyond the general header regardless of the topic specified in the rule if the rule has specified a role?

Reply:
I don't think this is the right approach. I think we should understand all APIkeys which do not have topic fields and then not perform the topic match on those keys. We should not rely on `req.request == nil` representing that group of APIkeys. There may be APIKeys which the kafka parser does not recognise beyond the generic header but still have a topic field.

Reply:
The intent is to allow  requests which have no topic field like apiVersions,  but are still specified with topic. (This will be the case when Role is expanded to include all non topic requests as well)

Reply:
I will work on restructuring this as per non-topic requests. Coming Soon!

Reply:
done. Please check if this looks fine